src/framework-ondata.js

```diff
@@ -1,18 +1,301 @@
+     vni.inputStates( portName, socketIndex, inputState );
+
+     if ( shouldRunUpdater( vni ) ) { 
+
+         // Save vars we will need in the promise where the context is different
+         var outputPorts = this.nodeInstance.outPorts;
+
+         if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+         } else { 
+             // TODO: revisit this logic later, per rdf-pipeline/noflo-rdf-pipeline#35
+
+             // save data used to see if we have a different state after running updater
+             var lastErrorState = _.clone( vni.errorState() );  // Shallow clone 
+             var lastOutputLm = vni.outputState().lm;
+ 
+             clearStateData( vni.errorState ); // clear last error state
+
+             // Save wrapper to make it accessible from the Promise
+             var wrapper = this.nodeInstance.wrapper;
+
+             new Promise(function( resolve ) { 
+                 // Execute fRunUpdater which will also execute the updater
+                 resolve( wrapper.fRunUpdater( vni ) );
+
+             }).then( function() { 
+                 // fRunUpdater/Updater success path
+             
+                 // Returned OK, but updater could have set an error - check for that
+                 if ( stateChange( vni.errorState, lastErrorState ) ) {
+                    setOutputErrorFlag( vni );   
+                    lastOutputLm = undefined;  // got a new error so we will force sending output
+                 }
+
+                 handleOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                 handleError( vni, outputPorts.error,  lastErrorState.lm );
+
+             }, function( rejected ) { 
+                 // fRunUpdater/updater failed 
+                 if ( isInitState( vni.errorState ) ) { 
+                     // fRunUpdater/updater has not already set an error state - use the rejected info
+                     changeStateData( vni.errorState, rejected );
+                 }
+
+                 setOutputErrorFlag( vni, true ); 
+
+                 stateChange( vni.outputState, lastOutputLm );
+                 stateChange( vni.errorState, lastErrorState );
+
+                 handleOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                 handleError( vni, outputPorts.error, lastErrorState.lm );
+                  
+
+                 // If we haven't already processed the rejected error, do it now
+                 if ( rejected !== vni.errorState().data ) { 
+                     lastErrorState = _.clone( vni.errorState() );   
+                     changeStateData( vni.errorState, rejected );
+                     if ( changeStateData( vni.errorState, lastErrorState ) ) { 
```

Comment:
The function changeStateData has no return statement, so this condition would never be fulfilled.